### PR TITLE
Syncing optimizations

### PR DESCRIPTION
**Pull request by mfrankovi**
Moving queueing syncing blocks from pbft to networking
Improvements to validation of syncing
Optimizing syncing to avoid data copy and calculating dag block order
Adding hash for dag order and unique transactions order in pbf block

### CODE REVIEW

File: src/consensus/pbft_manager.cpp [hunk 1]
```diff
@@ -1271,9 +1271,51 @@ std::pair<blk_hash_t, bool> PbftManager::proposeMyPbftBlock_() {
 
   uint64_t propose_pbft_period = pbft_chain_->getPbftChainSize() + 1;
   addr_t beneficiary = node_addr_;
+
+  // get DAG block and transaction order
+  auto dag_block_order = dag_mgr_->getDagBlockOrder(dag_block_hash);
+  if (dag_block_order.second->empty()) {
+    LOG(log_er_) << "DAG anchor block hash " << dag_block_hash << " getDagBlockOrder failed in propose";
+    assert(false);
+    return std::make_pair(NULL_BLOCK_HASH, true);
```
Comment (JakubFornadel) [inline src/consensus/pbft_manager.cpp:1280]: what is the logic behind return true on failures ? On line 1220 we return false for example...

File: src/consensus/pbft_manager.cpp [hunk 2]
```diff
@@ -1590,50 +1614,70 @@ bool PbftManager::pushPbftBlock_(PbftBlockCert const &pbft_block_cert_votes, vec
...
+  if (!sync) {
+    std::unordered_set<trx_hash_t> trx_set;
+    std::vector<trx_hash_t> transactionsToQuery;
+    DbStorage::MultiGetQuery db_query(db_);
+    db_query.append(DbStorage::Columns::dag_blocks, dag_blocks_order);
+    auto dag_blocks_res = db_query.execute();
+
+    for (auto const &dag_blk_raw : dag_blocks_res) {
+      DagBlock dag_block(asBytes(dag_blk_raw));
+      for (auto const trx_hash : dag_block.getTrxs()) {
```
Comment (JakubFornadel) [inline src/consensus/pbft_manager.cpp:1665]: maybe const auto& trx_hash ?

File: src/dag/dag.cpp [hunk 3]
```diff
@@ -293,7 +295,8 @@ DagManager::DagManager(blk_hash_t const &genesis, addr_t node_addr, std::shared_
       db_(db),
       anchor_(genesis),
       period_(0),
-      genesis_(genesis) {
+      genesis_(genesis),
+      log_time_(log_time) {
```
Comment (JakubFornadel) [inline src/dag/dag.cpp:299]: what is this log_time good for ?

File: src/dag/dag.cpp [hunk 4]
```diff
@@ -320,6 +323,8 @@ std::shared_ptr<DagManager> DagManager::getShared() {
 void DagManager::stop() {
   unique_lock lock(mutex_);
   trx_mgr_ = nullptr;
+  stopped_ = true;
+  block_worker_.join();
```
Comment (JakubFornadel) [inline src/dag/dag.cpp:327]: I think this should be: `if (block_worker_.joinable()) {  block_worker_.join(); }` + you should call stop() function in destructor as you are handling new thread and without explicitly calling `stop()` function (lets say in test) there would be a crash that might be hard to debug.

File: src/dag/dag.hpp [hunk 5]
```diff
@@ -122,10 +122,12 @@ class DagManager : public std::enable_shared_from_this<DagManager> {
 
   explicit DagManager(blk_hash_t const &genesis, addr_t node_addr, std::shared_ptr<TransactionManager> trx_mgr,
                       std::shared_ptr<PbftChain> pbft_chain, std::shared_ptr<DagBlockManager> dag_blk_mgr,
-                      std::shared_ptr<DbStorage> db);
+                      std::shared_ptr<DbStorage> db, logger::Logger log_time);
   virtual ~DagManager() = default;
   std::shared_ptr<DagManager> getShared();
+  void start();
   void stop();
+  void setNetwork(std::weak_ptr<Network> network) { network_ = move(network); }
```
Comment (JakubFornadel) [inline src/dag/dag.hpp:130]: let's use namespaces as it should be used so `move -> std::move`

File: src/dag/dag.hpp [hunk 6]
```diff
@@ -194,14 +194,18 @@ class DagManager : public std::enable_shared_from_this<DagManager> {
...
   blk_hash_t old_anchor_;  // anchor of the second to last period
   uint64_t period_;        // last period
   blk_hash_t genesis_;
   std::map<uint64_t, std::vector<blk_hash_t>> non_finalized_blks_;
-  std::map<uint64_t, std::vector<blk_hash_t>> finalized_blks_;
   DagFrontier frontier_;
+  std::atomic<bool> stopped_ = true;
+  std::thread block_worker_;
+
+  logger::Logger log_time_;
```
Comment (JakubFornadel) [inline src/dag/dag.hpp:208]: Cannot we use standard logger ?

File: src/network/network.hpp [hunk 7]
```diff
@@ -53,6 +53,14 @@ class Network {
   bool pbft_syncing();
   uint64_t syncTimeSeconds() const;
 
+  uint64_t pbftSyncingPeriod() const;
+  void syncBlockQueuePop();
+  std::shared_ptr<std::pair<SyncBlock, dev::p2p::NodeID>> processSyncBlock();
```
Comment (JakubFornadel) [inline src/network/network.hpp:58]: Why does it return shared_ptr ? I dont see any copies that would justify using pointer. If it is here to check if there is some return value, let's use std::optional

File: src/network/taraxa_capability.hpp [hunk 8]
```diff
@@ -146,6 +156,7 @@ struct TaraxaCapability : virtual CapabilityFace {
   util::ThreadPool syncing_tp_{1, false};
   SyncingState syncing_state_;
   PeersState peers_state_;
+  mutable std::shared_mutex sync_access_;
```
Comment (JakubFornadel) [inline src/network/taraxa_capability.hpp:159]: I would not know what is this mutex guarding without some checking. On the other hand if you group it with sync_queue_ (just move it's declaration next to it and maybe give some empty new line around) it would be clear straight away

File: src/network/taraxa_capability.cpp [hunk 9]
```diff
@@ -1612,6 +1593,93 @@ void TaraxaCapability::broadcastPreviousRoundNextVotesBundle() {
...
+
+void TaraxaCapability::syncBlockQueuePop() {
+  unique_lock lock(sync_access_);
+  sync_queue_.pop();
+}
+
+void TaraxaCapability::handleMaliciousSyncBlock(NodeID const &id) {
+  syncing_state_.set_peer_malicious(id);
+  auto host = host_.lock();
+  if (host) host->disconnect(sync_queue_.front().second, p2p::UserReason);
```
Comment (JakubFornadel) [inline src/network/taraxa_capability.cpp:1613]: shouldn't you use `id` instead of `sync_queue_.front().second` ?

File: src/network/network.hpp [hunk 10]
```diff
@@ -53,6 +53,14 @@ class Network {
...
+  void syncBlockQueuePush(SyncBlock const &block);
+  void clearSyncBlockQueue();
+  size_t syncBlockQueueSize() const;
+  void handleMaliciousSyncBlock(NodeID const &id);
```
Comment (JakubFornadel) [inline src/network/network.hpp:62]: So I went through the changes and I can see you moved `sync_queue_` from `pbft_manager` to `taraxa_capability` but I cant really see any added value in a way it is done (sure I might be wrong but this is how I understand the changes from quick review). Let me explain:

- I still think that the ultimate goal is to remove processing of sync blocks from pbft_manager and move it all somehow to tarcap. Then during syncing pbft_manager could be sort of "disabled"... 
- What we have here is that all the processing of sync blocks is still inside pbft_manager and as a result of moving `sync_queue_` into the tarcap is that you had to create 7 new methods inside network class as well as tarcap class, which is a lot of methods. 
- tarcap in fact does not really use `sync_queue_` except pushing blocks into it.  Only pbft_manager depends on it and use it internally. So if you leave the queue inside pbft_manager you can delete those 14 methods. I would keep maybe only adjusted `handleMaliciousSyncBlock()` function inside tarcap that would set malicious flag and disconnect the peer...

In summary, as we did not  remove also actual processing of sync blocks from pbft_manager, these changes seem like it complicated stuff rather then improved it ?

File: src/consensus/pbft_chain.cpp [hunk 11]
```diff
@@ -14,18 +14,22 @@ PbftBlock::PbftBlock(bytes const& b) : PbftBlock(dev::RLP(b)) {}
 PbftBlock::PbftBlock(dev::RLP const& r) {
   dev::RLP const rlp(r);
   if (!rlp.isList()) throw std::invalid_argument("PBFT RLP must be a list");
-  prev_block_hash_ = rlp[0].toHash<blk_hash_t>();
-  dag_block_hash_as_pivot_ = rlp[1].toHash<blk_hash_t>();
-  period_ = rlp[2].toInt<uint64_t>();
-  timestamp_ = rlp[3].toInt<uint64_t>();
-  signature_ = rlp[4].toHash<sig_t>();
```
Comment (MatusKysel) [inline src/consensus/pbft_chain.cpp:21]: btw no need to change that, if you accessing RLP elements in ascending order it is exactly same as iterator

File: src/consensus/pbft_manager.cpp [hunk 12]
```diff
@@ -1271,9 +1271,51 @@ std::pair<blk_hash_t, bool> PbftManager::proposeMyPbftBlock_() {
...
+    auto dag_blk = dag_blk_mgr_->getDagBlock(blk_hash);
+    if (!dag_blk) {
+      LOG(log_er_) << "DAG anchor block hash " << dag_block_hash << " getDagBlock failed in propose for block "
+                   << blk_hash;
+      assert(false);
+      return std::make_pair(NULL_BLOCK_HASH, true);
+    }
+    order_stream << dag_blk->getHash();
+    std::vector<trx_hash_t> trx_hashes;
+    for (auto const &trx_hash : dag_blk->getTrxs()) {
```
Comment (MatusKysel) [inline src/consensus/pbft_manager.cpp:1295]: why you looping this, when it returns vector of trx that you need

File: src/consensus/pbft_manager.cpp [hunk 13]
```diff
@@ -1271,9 +1271,51 @@ std::pair<blk_hash_t, bool> PbftManager::proposeMyPbftBlock_() {
...
+        assert(false);
+      }
+    }
+  }
+  order_stream.appendList(non_executed_transactions.size());
+  for (auto const &trx_hash : non_executed_transactions) {
+    order_stream << trx_hash;
+  }
+
+  auto order_hash = dev::sha3(order_stream.out());
```
Comment (MatusKysel) [inline src/consensus/pbft_manager.cpp:1314]: this whole section could be separate function

File: src/consensus/pbft_manager.cpp [hunk 14]
```diff
@@ -1590,50 +1614,70 @@ bool PbftManager::pushPbftBlock_(PbftBlockCert const &pbft_block_cert_votes, vec
...
+  auto const &anchor_hash = sync_block.pbft_blk->getPivotDagBlockHash();
+  dag_mgr_->setDagBlockOrder(anchor_hash, pbft_period, dag_blocks_order, batch);
 
-  std::vector<Transaction> transactions;
-  transactions.reserve(transactions_res.size());
-  for (auto const &trx_raw : transactions_res) {
-    if (trx_raw.size() > 0) transactions.emplace_back(asBytes(trx_raw));
+  if (!sync) {
+    std::unordered_set<trx_hash_t> trx_set;
+    std::vector<trx_hash_t> transactionsToQuery;
```
Comment (MatusKysel) [inline src/consensus/pbft_manager.cpp:1658]: naming

File: src/storage/db_storage.cpp [hunk 15]
```diff
@@ -251,6 +251,30 @@ std::vector<std::shared_ptr<DagBlock>> DbStorage::getDagBlocksAtLevel(level_t le
...
+void DbStorage::updateDagBlockCounters(Batch& write_batch, vector<DagBlock> blks) {
+  // Lock is needed since we are editing some fields
+  lock_guard<mutex> u_lock(dag_blocks_mutex_);
+  for (auto const& blk : blks) {
+    auto level = blk.getLevel();
+    std::string blocks = getBlocksByLevel(level);
+    if (blocks == "") {
+      blocks = blk.getHash().toString();
+    } else {
+      blocks = blocks + "," + blk.getHash().toString();
```
Comment (MatusKysel) [inline src/storage/db_storage.cpp:263]: why we are storing it as string? isn't it better to have rlp array of block hashes or something?

File: src/consensus/pbft_chain.cpp (hunk 11 above)
Comment (mfrankovi) [reply src/consensus/pbft_chain.cpp:21]: Internally RLP does not have random access, so accessing by index actually internally makes it iterate multiple time

Comment (MatusKysel) [reply src/consensus/pbft_chain.cpp:21]: yes, but it is not true if you doing it in ascending order, because it is doing same thing as you, https://github.com/Taraxa-project/taraxa-aleth/blob/taraxa_master/libdevcore/RLP.cpp#L66 you can see, it has internal iterator `m_lastEnd`, so no need to rewrite anything if your are accessing something like 1,2,3... or even 1,4,12 ...

File: src/consensus/pbft_manager.cpp (hunk 1 above)
Comment (mfrankovi) [reply src/consensus/pbft_manager.cpp:1280]: I have removed the return after assert but I also do not understand the meaning of the flag in the return values of this method. Lets talk about this

File: src/dag/dag.cpp (hunk 3 above)
Comment (mfrankovi) [reply src/dag/dag.cpp:299]: this log_time is some performance logger used across different classes. I am not sure if we actually use it for something. We need to check with everyone and remove all instances if it is not used

File: src/storage/db_storage.cpp (hunk 15 above)
Comment (mfrankovi) [reply src/storage/db_storage.cpp:263]: Yes, this should be modified. I would do it later as a separate PR and change all the places we use this.

File: src/consensus/pbft_manager.cpp [hunk 16]
```diff
@@ -1271,9 +1271,51 @@ std::pair<blk_hash_t, bool> PbftManager::proposeMyPbftBlock_() {
...
+    order_stream << dag_blk->getHash();
+    std::vector<trx_hash_t> trx_hashes;
+    for (auto const &trx_hash : dag_blk->getTrxs()) {
+      trx_hashes.emplace_back(trx_hash);
+    }
+    auto trx_statuses = db_->getTransactionStatus(trx_hashes);
+    for (uint32_t i = 0; i < trx_statuses.size(); i++) {
+      if (trx_statuses[i].state == TransactionStatusEnum::in_block) {
+        non_executed_transactions.emplace_back(trx_hashes[i]);
+      } else if (trx_statuses[i].state != TransactionStatusEnum::executed) {
```
[the remark this answers is not in the document]
Comment (mfrankovi) [reply src/consensus/pbft_manager.cpp:1302]: I am not sure I udnerstand this race condition. Lets talk about this

File: src/consensus/pbft_chain.cpp (hunk 11 above)
Comment (mfrankovi) [reply src/consensus/pbft_chain.cpp:21]: I see. Thanks

Comment (MatusKysel) [reply src/consensus/pbft_chain.cpp:21]: you do not need to change it, just letting you know

File: src/network/network.hpp (hunk 10 above)
Comment (mfrankovi) [reply src/network/network.hpp]: I have made some additional changes after your comment. There was some leftover processing in pbft manager that I also moved to tarcap. Network and tarcap have this issue where this maybe should be a single class since it is unlikely that we would have other capabilities. Currently only 4 methods are exposed in network that deal with the queue.

Comment (JakubFornadel) [reply src/network/network.hpp]: I dont see any changes that would move sync pbft blocks processing from pbft_manager to tarcap, so the comment from above still holds

File: src/network/sync_block.hpp [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+struct Transaction;
+class DagBlock;
+
+class SyncBlock {
+ public:
+  SyncBlock(PbftBlock const& pbft_blk, std::vector<Vote> const& cert_votes);
+  SyncBlock(dev::RLP const& all_rlp);
+  SyncBlock(bytes const& all_rlp);
+
+  std::shared_ptr<PbftBlock> pbft_blk;
```
Comment (JakubFornadel) [inline src/network/sync_block.hpp:24]: underscore at the end is used for private class members
